### PR TITLE
Use the FileIO submodule in Spark writers and readers.

### PR DESCRIPTION
**Pull request by mccheah**
Tricky because the table operations needs to be exposed. Added a mixed interface of Table + TableOperations accordingly. Therefore now the Spark data source is opinionated about ensuring the returned table also implements `HasTableOperations`.

### CODE REVIEW

File: spark/src/main/java/com/netflix/iceberg/spark/source/IcebergSource.java [hunk 1]
```diff
@@ -89,10 +90,10 @@ public DataSourceReader createReader(DataSourceOptions options) {
           .toUpperCase(Locale.ENGLISH));
     }
 
-    return Optional.of(new Writer(table, lazyConf(), format));
+    return Optional.of(new Writer(table, format, table.operations().io()));
   }
 
-  protected Table findTable(DataSourceOptions options) {
+  protected TableWithTableOperations findTable(DataSourceOptions options) {
```
Comment (mccheah) [inline spark/src/main/java/com/netflix/iceberg/spark/source/IcebergSource.java]: @rdblue this is the most interesting question raised by this patch. What do you think?

Comment (rdblue) [reply spark/src/main/java/com/netflix/iceberg/spark/source/IcebergSource.java]: What about adding `io` to the `Table` interface? I'd rather do that since `FileIO` is a public interface. I think that is mostly what HasTableOperations is used for anyway.

Comment (mccheah) [reply spark/src/main/java/com/netflix/iceberg/spark/source/IcebergSource.java]: I think that's fine.

File: api/src/main/java/com/netflix/iceberg/Table.java [hunk 2]
```diff
@@ -171,4 +172,10 @@ default AppendFiles newFastAppend() {
    * @return a new {@link Transaction}
    */
   Transaction newTransaction();
+
+  /**
+   * @return a {@link FileIO} to read and write table data and metadata files
+   */
+  FileIO io();
```
Comment (mccheah) [inline api/src/main/java/com/netflix/iceberg/Table.java]: Should `TableOperations#io` continue to exist?

Comment (rdblue) [reply api/src/main/java/com/netflix/iceberg/Table.java]: Yes, because `TableOperations` is still how implementations are passed.

File: spark/src/main/java/com/netflix/iceberg/spark/source/Writer.java [hunk 3]
```diff
@@ -235,21 +224,23 @@ public String toString() {
...
           Preconditions.checkNotNull(objectStore,
               "Cannot use object storage, missing location: " + OBJECT_STORE_PATH);
-          Path objectStorePath = new Path(objectStore);
 
           outputPathFunc = key -> {
             String partitionAndFilename = key.toPath() + "/" + filename;
             int hash = HASH_FUNC.apply(partitionAndFilename);
-            return new Path(objectStorePath,
-                String.format("%08x/%s/%s", hash, context, partitionAndFilename));
+            return UriBuilder.fromUri(URI.create(objectStore))
```
Comment (rdblue) [inline spark/src/main/java/com/netflix/iceberg/spark/source/Writer.java]: The Path/URI methods cause problems with escape characters. This should use strings for reliability.

File: spark/src/test/java/com/netflix/iceberg/spark/source/TestTables.java [hunk 4]
```diff
@@ -73,7 +72,8 @@ private TestTable(TestTableOperations ops, String name) {
       this.ops = ops;
     }
 
-    TestTableOperations ops() {
+    @Override
+    public TestTableOperations operations() {
```
Comment (rdblue) [inline spark/src/test/java/com/netflix/iceberg/spark/source/TestTables.java]: Why was this rename needed?

**Comment by rdblue** [comment]
@mccheah, thanks for working on this! It's about ready to go, but we need to change back to using String instead of URI. There's also an unnecessary rename that we should remove.

File: spark/src/test/java/com/netflix/iceberg/spark/source/TestTables.java (hunk 4 above)
Comment (mccheah) [reply spark/src/test/java/com/netflix/iceberg/spark/source/TestTables.java]: The previous version didn't actually override the superclass's method.

Comment (rdblue) [reply spark/src/test/java/com/netflix/iceberg/spark/source/TestTables.java]: Probably because we didn't expose `TableOperations` at first. Now that `FileIO` will be public, we can probably remove `HasTableOperations` in a follow-up.

File: spark/src/main/java/com/netflix/iceberg/spark/source/Writer.java [hunk 5]
```diff
@@ -202,31 +187,31 @@ public String toString() {
...
+        return new UnpartitionedWriter(dataLocation, filename, format, factory, fileIo);
       } else {
-        Path baseDataPath = lazyDataPath(); // avoid calling this in the output path function
-        Function<PartitionKey, Path> outputPathFunc = key ->
-            new Path(new Path(baseDataPath, key.toPath()), filename);
+        String baseDataPath = stripTrailingSlash(dataLocation); // avoid calling this in the output path function
+        Function<PartitionKey, String> outputPathFunc = key ->
+            String.format("%s/%s/%s",
+                stripTrailingSlash(baseDataPath),
+                stripTrailingSlash(stripLeadingSlash(key.toPath())),
```
Comment (rdblue) [inline spark/src/main/java/com/netflix/iceberg/spark/source/Writer.java]: This is always implemented by [`PartitionSpec#partitionToPath`](https://github.com/apache/incubator-iceberg/blob/master/api/src/main/java/com/netflix/iceberg/PartitionSpec.java#L143-L156). That never adds a starting or trailing `/` so this can rely on that and doesn't need to do the extra check.

File: spark/src/main/java/com/netflix/iceberg/spark/source/Writer.java [hunk 6]
```diff
@@ -202,31 +187,31 @@ public String toString() {
...
-
+        return new UnpartitionedWriter(dataLocation, filename, format, factory, fileIo);
       } else {
-        Path baseDataPath = lazyDataPath(); // avoid calling this in the output path function
-        Function<PartitionKey, Path> outputPathFunc = key ->
-            new Path(new Path(baseDataPath, key.toPath()), filename);
+        String baseDataPath = stripTrailingSlash(dataLocation); // avoid calling this in the output path function
+        Function<PartitionKey, String> outputPathFunc = key ->
+            String.format("%s/%s/%s",
+                stripTrailingSlash(baseDataPath),
```
Comment (rdblue) [inline spark/src/main/java/com/netflix/iceberg/spark/source/Writer.java]: `baseDataPath` already stripped the trailing `/`.

File: spark/src/main/java/com/netflix/iceberg/spark/source/Writer.java [hunk 7]
```diff
@@ -202,31 +187,31 @@ public String toString() {
...
       AppenderFactory<InternalRow> factory = new SparkAppenderFactory();
       if (spec.fields().isEmpty()) {
-        return new UnpartitionedWriter(lazyDataPath(), filename, format, conf.value(), factory);
-
+        return new UnpartitionedWriter(dataLocation, filename, format, factory, fileIo);
       } else {
-        Path baseDataPath = lazyDataPath(); // avoid calling this in the output path function
-        Function<PartitionKey, Path> outputPathFunc = key ->
-            new Path(new Path(baseDataPath, key.toPath()), filename);
+        String baseDataPath = stripTrailingSlash(dataLocation); // avoid calling this in the output path function
```
Comment (rdblue) [inline spark/src/main/java/com/netflix/iceberg/spark/source/Writer.java]: Since this is an internal implementation, I think that the `dataLocation()` method should guarantee that the trailing `/` is removed. That way we never need to do this in the tasks.

File: spark/src/main/java/com/netflix/iceberg/spark/source/Writer.java [hunk 8]
```diff
@@ -235,21 +220,25 @@ public String toString() {
...
 
           outputPathFunc = key -> {
-            String partitionAndFilename = key.toPath() + "/" + filename;
+            String partitionAndFilename = String.format("%s/%s", key.toPath(), filename);
             int hash = HASH_FUNC.apply(partitionAndFilename);
-            return new Path(objectStorePath,
-                String.format("%08x/%s/%s", hash, context, partitionAndFilename));
+            return String.format(
+                "%s/%08x/%s/%s/%s",
+                stripTrailingSlash(objectStore),
```
Comment (rdblue) [inline spark/src/main/java/com/netflix/iceberg/spark/source/Writer.java]: Same general ideas here as above. This shouldn't remove slashes if it can be done in one central place or strings are generated by Iceberg code. We may need to document that other methods must not produce trailing or starting `/`, but that is a better way to go.

**Comment by rdblue** [comment]
@mccheah, I think this is just about ready to merge, but it doesn't need most of the calls to strip slashes. Once those are fixed, I'll merge it. No rush since you're probably just back from the holidays.

**Comment by mccheah** [comment]
Almost there, addressed the comments, fixed the merge conflicts.

**Comment by rdblue** [comment]
Merged. Thanks @mccheah!
